// ts-simple-ast/dist-scripts/src/tests/compiler/type/typeTests.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var typescript_1 = require("../../../typescript");
var fileSystem_1 = require("../../../fileSystem");
var testHelpers_1 = require("../testHelpers");
describe("Type", function () {
    function getInfoFromTextWithTypeChecking(text) {
        return testHelpers_1.getInfoFromText(text, { host: new fileSystem_1.VirtualFileSystemHost(), includeLibDts: true, compilerOptions: { strictNullChecks: true } });
    }
    function getTypeFromText(text) {
        var result = getInfoFromTextWithTypeChecking(text);
        return __assign({}, result, { firstType: result.firstChild.getDeclarations()[0].getType() });
    }
    function getTypeAliasTypeFromText(text) {
        var result = getInfoFromTextWithTypeChecking(text);
        return __assign({}, result, { firstType: result.firstChild.getTypeNodeOrThrow().getType() });
    }
    // todo: move all the tests to happen in here because this will speed up the tests quite a lot
    describe("fast test", function () {
        var text = "\nenum EmptyEnum { }\nenum MyEnum { value, value2 }\ninterface MyInterface {}\nclass MyClass {}\n\nlet anonymousType: { str: string; };\nlet stringType: string;\nlet booleanType: boolean;\nlet numberType: number;\nlet booleanLiteralType: true;\nlet numberLiteralType: 5;\nlet stringLiteralType: 'test';\nlet enumType: EmptyEnum; // no clue why, but this is only an enum type for empty enums\nlet enumLiteralType: MyEnum.value;\nlet interfaceType: MyInterface;\nlet intersectionType: string & number;\nlet unionType: string | number;\nlet objectType: { prop: string; };\nlet tupleType: [string];\nlet tupleTypeMultiple: [string, number];\nlet genericArrayType: Array<string>;\nlet arrayType: string[];\nlet arrayTypeOfTuples: [string][];\nlet undefinedType: undefined;\nlet classType: MyClass;\nlet functionType: () => string;\nlet constructorType: { new(): MyClass; };\nlet indexedType: { [index: string]: object; [index: number]: Date; }\nlet emptyObjectType: { };\nlet stringWithUndefinedType: string | undefined;\nlet stringWithNullType: string | null;\nlet stringWithUndefinedAndNullType: string | undefined | null;\n";
        var sourceFile = getInfoFromTextWithTypeChecking(text).sourceFile;
        var typesByName = {};
        try {
            for (var _a = __values(sourceFile.getVariableDeclarations()), _b = _a.next(); !_b.done; _b = _a.next()) {
                var dec = _b.value;
                typesByName[dec.getName()] = dec.getType();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_1) throw e_1.error; }
        }
        describe("compilerType", function () {
            it("should get the compiler type", function () {
                chai_1.expect(typesByName["stringType"].compilerType.flags).to.equal(typescript_1.TypeFlags.String);
            });
        });
        describe("getUnionTypes", function () {
            it("should get them when there aren't any", function () {
                chai_1.expect(typesByName["stringType"].getUnionTypes().length).to.equal(0);
            });
            it("should get them when they exist", function () {
                var firstType = typesByName["unionType"];
                chai_1.expect(firstType.getUnionTypes().length).to.equal(2);
                chai_1.expect(firstType.getUnionTypes()[0].getFlags()).to.equal(typescript_1.TypeFlags.String);
                chai_1.expect(firstType.getUnionTypes()[1].getFlags()).to.equal(typescript_1.TypeFlags.Number);
            });
            it("should not return anything for an intersection type", function () {
                chai_1.expect(typesByName["intersectionType"].getUnionTypes().length).to.equal(0);
            });
        });
        describe("getIntersectionTypes", function () {
            it("should get them when there aren't any", function () {
                chai_1.expect(typesByName["stringType"].getIntersectionTypes().length).to.equal(0);
            });
            it("should get them when they exist", function () {
                var firstType = typesByName["intersectionType"];
                chai_1.expect(firstType.getIntersectionTypes().length).to.equal(2);
                chai_1.expect(firstType.getIntersectionTypes()[0].getFlags()).to.equal(typescript_1.TypeFlags.String);
                chai_1.expect(firstType.getIntersectionTypes()[1].getFlags()).to.equal(typescript_1.TypeFlags.Number);
            });
            it("should not return anything for a union type", function () {
                chai_1.expect(typesByName["unionType"].getIntersectionTypes().length).to.equal(0);
            });
        });
        describe("isAnonymousType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isAnonymousType()).to.equal(expected);
            }
            it("should get when it is an anonymous type", function () {
                doTest("anonymousType", true);
            });
            it("should get when it's not an anonymous type", function () {
                doTest("stringType", false);
            });
        });
        describe("isBooleanType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isBooleanType()).to.equal(expected);
            }
            it("should get when it is a boolean type", function () {
                doTest("booleanType", true);
            });
            it("should get when it's not a boolean type", function () {
                doTest("stringType", false);
            });
        });
        describe("isStringType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isStringType()).to.equal(expected);
            }
            it("should get when it is a string type", function () {
                doTest("stringType", true);
            });
            it("should get when it's not a string type", function () {
                doTest("booleanType", false);
            });
        });
        describe("isNumberType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isNumberType()).to.equal(expected);
            }
            it("should get when it is a number type", function () {
                doTest("numberType", true);
            });
            it("should get when it's not a number type", function () {
                doTest("booleanType", false);
            });
        });
        describe("isLiteralType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isLiteralType()).to.equal(expected);
            }
            it("should get when it is a boolean literal", function () {
                doTest("booleanLiteralType", true);
            });
            it("should get when it is a string literal", function () {
                doTest("stringLiteralType", true);
            });
            it("should get when it is an enum literal", function () {
                doTest("enumLiteralType", true);
            });
            it("should get when it is a number literal", function () {
                doTest("numberLiteralType", true);
            });
            it("should get when it's not", function () {
                doTest("booleanType", false);
            });
        });
        describe("isBooleanLiteralType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isBooleanLiteralType()).to.equal(expected);
            }
            it("should get when it is", function () {
                doTest("booleanLiteralType", true);
            });
            it("should get when it's not", function () {
                doTest("booleanType", false);
            });
        });
        describe("isEnumLiteralType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isEnumLiteralType()).to.equal(expected);
            }
            it("should get when it is", function () {
                doTest("enumLiteralType", true);
            });
            it("should get when it's not", function () {
                doTest("enumType", false);
            });
        });
        describe("isStringLiteralType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isStringLiteralType()).to.equal(expected);
            }
            it("should get when it is", function () {
                doTest("stringLiteralType", true);
            });
            it("should get when it's not", function () {
                doTest("stringType", false);
            });
        });
        describe("isNumberLiteralType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isNumberLiteralType()).to.equal(expected);
            }
            it("should get when it is", function () {
                doTest("numberLiteralType", true);
            });
            it("should get when it's not", function () {
                doTest("numberType", false);
            });
        });
        describe("isEnumType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isEnumType()).to.equal(expected);
            }
            it("should get when it is an enum type", function () {
                doTest("enumType", true);
            });
            it("should get when it's not an enum type", function () {
                doTest("enumLiteralType", false);
            });
        });
        describe("isInterfaceType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isInterfaceType()).to.equal(expected);
            }
            it("should get when it is", function () {
                doTest("interfaceType", true);
            });
            it("should get when it's not", function () {
                doTest("stringType", false);
            });
        });
        describe("isIntersectionType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isIntersectionType()).to.equal(expected);
            }
            it("should get when it is an intersection type", function () {
                doTest("intersectionType", true);
            });
            it("should get when it's not an intersection type", function () {
                doTest("unionType", false);
            });
        });
        describe("isUnionType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isUnionType()).to.equal(expected);
            }
            it("should get when it is a union type", function () {
                doTest("unionType", true);
            });
            it("should get when it's not a union type", function () {
                doTest("intersectionType", false);
            });
        });
        describe("isObjectType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isObjectType()).to.equal(expected);
            }
            it("should get when it is an object type", function () {
                doTest("objectType", true);
            });
            it("should get when it's not an object type", function () {
                doTest("numberType", false);
            });
        });
        describe("isTupleType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isTupleType()).to.equal(expected);
            }
            it("should be when tuple and one element", function () {
                doTest("tupleType", true);
            });
            it("should be when tuple and multiple", function () {
                doTest("tupleTypeMultiple", true);
            });
            it("should not be when not an array", function () {
                doTest("arrayType", false);
            });
            it("should not be when not an array of tuples", function () {
                doTest("arrayTypeOfTuples", false);
            });
            it("should not be when not tuple", function () {
                doTest("stringType", false);
            });
        });
        describe("isUndefinedType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isUndefinedType()).to.equal(expected);
            }
            it("should be when undefined", function () {
                doTest("undefinedType", true);
            });
            it("should not be when not undefined", function () {
                doTest("stringType", false);
            });
        });
        describe("getFlags", function () {
            it("should get the type flags", function () {
                chai_1.expect(typesByName["numberType"].getFlags()).to.equal(typescript_1.TypeFlags.Number);
            });
        });
        describe("getObjectFlags", function () {
            it("should get the object flags when not an object", function () {
                chai_1.expect(typesByName["numberType"].getObjectFlags()).to.equal(0);
            });
            it("should get the object flags when an object", function () {
                chai_1.expect(typesByName["interfaceType"].getObjectFlags()).to.equal(typescript_1.ObjectFlags.Interface);
            });
        });
        describe("getSymbol", function () {
            it("should get symbol when it has one", function () {
                chai_1.expect(typesByName["classType"].getSymbol().getName()).to.equal("MyClass");
            });
            it("should return undefined when it doesn't have one", function () {
                chai_1.expect(typesByName["stringType"].getSymbol()).to.be.undefined;
            });
        });
        describe("getSymbolOrThrow", function () {
            it("should get symbol when it has one", function () {
                chai_1.expect(typesByName["classType"].getSymbolOrThrow().getName()).to.equal("MyClass");
            });
            it("should return undefined when it doesn't have one", function () {
                chai_1.expect(function () { return typesByName["stringType"].getSymbolOrThrow(); }).to.throw();
            });
        });
        describe("getApparentType", function () {
            it("should get the apparent type", function () {
                chai_1.expect(typesByName["numberType"].getApparentType().getText()).to.equal("Number");
            });
        });
        describe("getCallSignatures", function () {
            it("should return no call signatures when none exist", function () {
                chai_1.expect(typesByName["stringType"].getCallSignatures().length).to.equal(0);
            });
            it("should return the call signatures of a type", function () {
                chai_1.expect(typesByName["functionType"].getCallSignatures().length).to.equal(1);
            });
        });
        describe("getConstructSignatures", function () {
            it("should return no construct signatures when none exist", function () {
                chai_1.expect(typesByName["stringType"].getConstructSignatures().length).to.equal(0);
            });
            it("should return the construct signatures of a type", function () {
                chai_1.expect(typesByName["constructorType"].getConstructSignatures().length).to.equal(1);
            });
        });
        describe("getStringIndexType", function () {
            it("should return undefined when no string index type", function () {
                chai_1.expect(typesByName["emptyObjectType"].getStringIndexType()).to.be.undefined;
            });
            it("should return the string index type", function () {
                chai_1.expect(typesByName["indexedType"].getStringIndexType().getText()).to.equal("object");
            });
        });
        describe("getNumberIndexType", function () {
            it("should return undefined when it doesn't have one", function () {
                chai_1.expect(typesByName["emptyObjectType"].getNumberIndexType()).to.be.undefined;
            });
            it("should return when it does", function () {
                chai_1.expect(typesByName["indexedType"].getNumberIndexType().getText()).to.equal("Date");
            });
        });
        describe("getNonNullableType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].getNonNullableType().getText()).to.equal(expected);
            }
            it("should return the original type for a type that's already non-nullable", function () {
                doTest("stringType", "string");
            });
            it("should return the non-nullable type for undefined", function () {
                doTest("stringWithUndefinedType", "string");
            });
            it("should return the non-nullable type for null", function () {
                doTest("stringWithNullType", "string");
            });
            it("should return the non-nullable type for null and undefined", function () {
                doTest("stringWithUndefinedAndNullType", "string");
            });
        });
        describe("isNullable", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].isNullable()).to.equal(expected);
            }
            it("should return false for a non-nullable type", function () {
                doTest("stringType", false);
            });
            it("should return true for undefined", function () {
                doTest("stringWithUndefinedType", true);
            });
            it("should return true for null", function () {
                doTest("stringWithNullType", true);
            });
            it("should return true for null and undefined", function () {
                doTest("stringWithUndefinedAndNullType", true);
            });
            it("should return true for an optional property", function () {
                var firstChild = getInfoFromTextWithTypeChecking("function test(param?: string) {}").firstChild;
                chai_1.expect(firstChild.getParameters()[0].getType().isNullable()).to.equal(true);
            });
        });
        describe("getTupleElements", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].getTupleElements().map(function (t) { return t.getText(); })).to.deep.equal(expected);
            }
            it("should get the tuple type's types", function () {
                doTest("tupleTypeMultiple", ["string", "number"]);
            });
            it("should get nothing when not a tuple", function () {
                doTest("stringType", []);
            });
            it("should get nothing for an array type that has type arguments", function () {
                doTest("genericArrayType", []);
            });
        });
        describe("getTypeArguments", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].getTypeArguments().map(function (t) { return t.getText(); })).to.deep.equal(expected);
            }
            it("should get the type arguments for an array", function () {
                doTest("genericArrayType", ["string"]);
            });
            it("should get them for a tuple", function () {
                doTest("tupleTypeMultiple", ["string", "number"]);
            });
            it("should get nothing when no type args", function () {
                doTest("stringType", []);
            });
        });
        describe("getBaseTypeOfLiteralType", function () {
            function doTest(typeName, expected) {
                chai_1.expect(typesByName[typeName].getBaseTypeOfLiteralType().getText()).to.equal(expected);
            }
            it("should get it for a number literal", function () {
                doTest("numberLiteralType", "number");
            });
            it("should get it for a boolean literal", function () {
                doTest("booleanLiteralType", "boolean");
            });
            it("should get it for a string literal", function () {
                doTest("stringLiteralType", "string");
            });
            it("should return the same type for a string type", function () {
                doTest("stringType", "string");
            });
            it("should return the same type for any other type", function () {
                doTest("interfaceType", "MyInterface");
            });
        });
        var e_1, _c;
    });
    describe("getText", function () {
        var longType = "string | number | Date | { reallyReallyLooooooonnnnnnnnnnnnggggggggggggggggNnnnnnnaaaaaaaaammmmmmmmmmmeeeeee: string; }";
        it("should get the text", function () {
            var firstType = getTypeFromText("let myType: string[];").firstType;
            chai_1.expect(firstType.getText()).to.equal("string[]");
        });
        it("should get the text when providing the enclosing node", function () {
            var _a = getTypeFromText("let myType: " + longType + ";"), firstChild = _a.firstChild, firstType = _a.firstType;
            chai_1.expect(firstType.getText(firstChild)).to.equal(longType);
        });
        it("should use the type format flags", function () {
            var _a = getTypeFromText("let myType: " + longType + ";"), firstChild = _a.firstChild, firstType = _a.firstType;
            chai_1.expect(firstType.getText(firstChild, typescript_1.TypeFormatFlags.None)).to.equal(longType.substring(0, 97) + "...");
        });
    });
    describe("getProperties", function () {
        it("should get the properties when there are none", function () {
            var firstType = getTypeFromText("let myType: {};").firstType;
            chai_1.expect(firstType.getProperties().length).to.equal(0);
        });
        it("should get the properties of a non-object type", function () {
            var firstType = getTypeFromText("let myType: 1;").firstType;
            chai_1.expect(firstType.getProperties().length).to.equal(6);
        });
        it("should get the properties when some exist", function () {
            var firstType = getTypeFromText("let myType: { str: string; };").firstType;
            var props = firstType.getProperties();
            chai_1.expect(props.length).to.equal(1);
            chai_1.expect(props[0].getName()).to.equal("str");
        });
    });
    describe("getProperty", function () {
        it("should get the property by name", function () {
            var firstType = getTypeFromText("let myType: { str: string; other: number; };").firstType;
            var prop = firstType.getProperty("other");
            chai_1.expect(prop.getName()).to.equal("other");
        });
        it("should get the property by function", function () {
            var firstType = getTypeFromText("let myType: { str: string; other: number; };").firstType;
            var prop = firstType.getProperty(function (p) { return p.getName() === "other"; });
            chai_1.expect(prop.getName()).to.equal("other");
        });
    });
    describe("getApparentProperties", function () {
        it("should return the apparent properties of a type", function () {
            var firstType = getTypeFromText("let myType: 1;").firstType;
            chai_1.expect(firstType.getApparentProperties().length).to.equal(6);
        });
    });
    describe("getApparentProperty", function () {
        it("should get the property by name", function () {
            var firstType = getTypeFromText("let myType: { str: string; other: number; };").firstType;
            var prop = firstType.getApparentProperty("other");
            chai_1.expect(prop.getName()).to.equal("other");
        });
        it("should get the property by function", function () {
            var firstType = getTypeFromText("let myType: { str: string; other: number; };").firstType;
            var prop = firstType.getApparentProperty(function (p) { return p.getName() === "other"; });
            chai_1.expect(prop.getName()).to.equal("other");
        });
    });
    describe("getConstraint", function () {
        function doTest(text, expected) {
            var firstType = getTypeAliasTypeFromText(text).firstType;
            if (expected == null)
                chai_1.expect(firstType.getConstraint()).to.be.undefined;
            else
                chai_1.expect(firstType.getConstraint().getText()).to.equal(expected);
        }
        it("should get the constraint when it exists", function () {
            doTest("type t<T extends string> = T;", "string");
        });
        it("should be undefined when it doesn't have a constraint", function () {
            doTest("type t<T> = T;", undefined);
        });
    });
    describe("getConstraintOrThrow", function () {
        function doTest(text, expected) {
            var firstType = getTypeAliasTypeFromText(text).firstType;
            if (expected == null)
                chai_1.expect(function () { return firstType.getConstraintOrThrow(); }).to.throw();
            else
                chai_1.expect(firstType.getConstraintOrThrow().getText()).to.equal(expected);
        }
        it("should get the constraint when it exists", function () {
            doTest("type t<T extends string> = T;", "string");
        });
        it("should be undefined when it doesn't have a constraint", function () {
            doTest("type t<T> = T;", undefined);
        });
    });
    describe("getDefault", function () {
        function doTest(text, expected) {
            var firstType = getTypeAliasTypeFromText(text).firstType;
            if (expected == null)
                chai_1.expect(firstType.getDefault()).to.be.undefined;
            else
                chai_1.expect(firstType.getDefault().getText()).to.equal(expected);
        }
        it("should get the default when it exists", function () {
            doTest("type t<T = string> = T;", "string");
        });
        it("should be undefined when it doesn't have a default", function () {
            doTest("type t<T> = T;", undefined);
        });
    });
    describe("getDefaultOrThrow", function () {
        function doTest(text, expected) {
            var firstType = getTypeAliasTypeFromText(text).firstType;
            if (expected == null)
                chai_1.expect(function () { return firstType.getDefaultOrThrow(); }).to.throw();
            else
                chai_1.expect(firstType.getDefaultOrThrow().getText()).to.equal(expected);
        }
        it("should get the default when it exists", function () {
            doTest("type t<T = string> = T;", "string");
        });
        it("should be undefined when it doesn't have a default", function () {
            doTest("type t<T> = T;", undefined);
        });
    });
    describe("getBaseTypes", function () {
        it("should return the base types of a type", function () {
            var firstType = getTypeFromText("let myType: MyInterface; interface MyInterface extends OtherInterface {}\ninterface OtherInterface").firstType;
            var baseTypes = firstType.getBaseTypes();
            chai_1.expect(baseTypes.length).to.equal(1);
            chai_1.expect(baseTypes[0].getText()).to.equal("OtherInterface");
        });
    });
    describe("getAliasSymbol", function () {
        it("should return the alias symbol when it exists", function () {
            var firstType = getTypeFromText("let myType: MyAlias; type MyAlias = {str: string;};").firstType;
            chai_1.expect(firstType.getAliasSymbol().getFlags()).to.equal(typescript_1.SymbolFlags.TypeAlias);
        });
        it("should return undefined when not exists", function () {
            var firstType = getTypeFromText("let myType: string;").firstType;
            chai_1.expect(firstType.getAliasSymbol()).to.be.undefined;
        });
    });
    describe("getAliasSymbolOrThrow", function () {
        it("should return the alias symbol when it exists", function () {
            var firstType = getTypeFromText("let myType: MyAlias; type MyAlias = {str: string;};").firstType;
            chai_1.expect(firstType.getAliasSymbolOrThrow().getFlags()).to.equal(typescript_1.SymbolFlags.TypeAlias);
        });
        it("should throw when not exists", function () {
            var firstType = getTypeFromText("let myType: string;").firstType;
            chai_1.expect(function () { return firstType.getAliasSymbolOrThrow(); }).to.throw();
        });
    });
    describe("getAliasTypeArguments", function () {
        it("should not have any when none exist", function () {
            var firstType = getTypeFromText("let myType: string;").firstType;
            chai_1.expect(firstType.getAliasTypeArguments().length).to.equal(0);
        });
        it("should return the type args when they exist", function () {
            var firstType = getTypeFromText("let myType: MyAlias<string>; type MyAlias<T> = {str: T;};").firstType;
            var typeArgs = firstType.getAliasTypeArguments();
            chai_1.expect(typeArgs.length).to.equal(1);
            chai_1.expect(typeArgs[0].getText()).to.equal("string");
        });
    });
});
