// ts-simple-ast/dist-scripts/src/tests/compiler/tools/languageServiceTests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var typescript_1 = require("../../../typescript");
var errors_1 = require("../../../errors");
var testHelpers_1 = require("../testHelpers");
describe("LanguageService", function () {
    describe("getEmitOutput", function () {
        it("should get the emit output", function () {
            function doTest(sourceFileOrFilePath) {
                var output = sourceFile.global.languageService.getEmitOutput(sourceFileOrFilePath);
                checkOutput(output, {
                    emitSkipped: false,
                    outputFiles: [{
                            fileName: "/" + sourceFile.getBaseName().replace(".ts", ".js"),
                            text: "var t = 5;\n",
                            writeByteOrderMark: false
                        }]
                });
            }
            var sourceFile = testHelpers_1.getInfoFromText("const t = 5;", { compilerOptions: { target: typescript_1.ScriptTarget.ES5 } }).sourceFile;
            doTest(sourceFile);
            doTest(sourceFile.getFilePath());
        });
        it("should get the emit output when specifying a source file", function () {
            var _a = testHelpers_1.getInfoFromText("const t = 5;", { compilerOptions: { target: typescript_1.ScriptTarget.ES5 } }), sourceFile = _a.sourceFile, project = _a.project;
            var output = sourceFile.global.languageService.getEmitOutput(sourceFile);
            checkOutput(output, {
                emitSkipped: false,
                outputFiles: [{
                        fileName: "/" + sourceFile.getBaseName().replace(".ts", ".js"),
                        text: "var t = 5;\n",
                        writeByteOrderMark: false
                    }]
            });
        });
        it("should only emit the declaration file when specified", function () {
            var _a = testHelpers_1.getInfoFromText("const t = 5;", { compilerOptions: { declaration: true } }), sourceFile = _a.sourceFile, project = _a.project;
            var output = sourceFile.global.languageService.getEmitOutput(sourceFile.getFilePath(), true);
            checkOutput(output, {
                emitSkipped: false,
                outputFiles: [{
                        fileName: "/" + sourceFile.getBaseName().replace(".ts", ".d.ts"),
                        text: "declare const t = 5;\n",
                        writeByteOrderMark: false
                    }]
            });
        });
        it("should not emit if there is a declaraton file error", function () {
            var _a = testHelpers_1.getInfoFromText("class MyClass {}\n export class Test extends MyClass {}\n", { compilerOptions: { declaration: true } }), sourceFile = _a.sourceFile, project = _a.project;
            var output = sourceFile.global.languageService.getEmitOutput(sourceFile.getFilePath(), true);
            checkOutput(output, {
                emitSkipped: true,
                outputFiles: [{
                        fileName: "/" + sourceFile.getBaseName().replace(".ts", ".d.ts"),
                        text: "export declare class Test extends MyClass {\n}\n",
                        writeByteOrderMark: false
                    }]
            });
        });
        it("should throw when the specified file does not exist", function () {
            var project = testHelpers_1.getInfoFromText("").project;
            chai_1.expect(function () { return project.getLanguageService().getEmitOutput("nonExistentFile.ts"); }).to.throw(errors_1.FileNotFoundError);
        });
    });
    describe("organizeImports", function () {
        it("should remove imports that don't exist", function () {
            var _a = testHelpers_1.getInfoFromText("import * as bravo from 'bravo';\nimport * as alpha from 'alpha';", { filePath: "/file.ts" }), sourceFile = _a.sourceFile, project = _a.project;
            var results = project.getLanguageService().organizeImports(sourceFile);
            chai_1.expect(results.length).to.equal(1);
            checkFileTextChanges(results[0], {
                fileName: "/file.ts",
                textChanges: [{
                        newText: "",
                        span: { start: 0, length: 32 }
                    }, {
                        newText: "",
                        span: { start: 32, length: 31 }
                    }]
            });
        });
        it("should organize imports when they're used", function () {
            var project = testHelpers_1.getInfoFromText("export default class MyClass {}", { filePath: "/MyClass.ts" }).project;
            project.createSourceFile("/MyInterface.ts", "export default interface MyInterface {}");
            project.createSourceFile("/UnusedInterface.ts", "export default interface Identifier {}");
            var sourceFile = project.createSourceFile("/main.ts", "import MyInterface from './MyInterface';\nimport MyClass from './MyClass';\n" +
                "import UnusedInterface from './UnusedInterface';\n" +
                "const myVar: MyInterface = new MyClass();");
            var results = project.getLanguageService().organizeImports(sourceFile.getFilePath());
            chai_1.expect(results.length).to.equal(1);
            checkFileTextChanges(results[0], {
                fileName: "/main.ts",
                textChanges: [{
                        newText: "import MyClass from './MyClass';\nimport MyInterface from './MyInterface';\n",
                        span: { start: 0, length: 41 }
                    }, {
                        newText: "",
                        span: { start: 41, length: 33 }
                    }, {
                        newText: "",
                        span: { start: 74, length: 49 }
                    }]
            });
        });
    });
});
function checkOutput(output, expected) {
    chai_1.expect(output.getEmitSkipped()).to.equal(expected.emitSkipped);
    chai_1.expect(output.getOutputFiles().length).to.equal(expected.outputFiles.length);
    for (var i = 0; i < expected.outputFiles.length; i++) {
        var actualFile = output.getOutputFiles()[i];
        var expectedFile = expected.outputFiles[i];
        chai_1.expect(actualFile.getFilePath()).to.equal(expectedFile.fileName, "fileName");
        chai_1.expect(actualFile.getText()).to.equal(expectedFile.text, "text");
        chai_1.expect(actualFile.getWriteByteOrderMark()).to.equal(expectedFile.writeByteOrderMark, "writeByteOrderMark");
    }
}
function checkFileTextChanges(actual, expected) {
    chai_1.expect(actual.getFilePath()).to.equal(expected.fileName, "fileName");
    chai_1.expect(actual.getTextChanges().length).to.equal(expected.textChanges.length, "textChangesLength");
    for (var i = 0; i < expected.textChanges.length; i++)
        checkTextChange(actual.getTextChanges()[i], expected.textChanges[i]);
}
function checkTextChange(actual, expected) {
    chai_1.expect(actual.getNewText()).to.equal(expected.newText, "newText");
    checkTextSpan(actual.getSpan(), expected.span);
}
function checkTextSpan(actual, expected) {
    chai_1.expect(actual.getStart()).to.equal(expected.start, "start");
    chai_1.expect(actual.getLength()).to.equal(expected.length, "length");
    chai_1.expect(actual.getEnd()).to.equal(expected.start + expected.length, "end");
}
