// ts-simple-ast/dist-scripts/src/tests/utils/tsconfig/getPathsFromTsConfigParseResultTests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var fileSystem_1 = require("../../../fileSystem");
var utils_1 = require("../../../utils");
describe("getPathsFromTsConfigParseResult", function () {
    function getFileAndDirPaths(fileSystem) {
        var fileSystemWrapper = new fileSystem_1.FileSystemWrapper(fileSystem);
        var tsConfigParseResult = utils_1.getTsConfigParseResult({
            tsConfigFilePath: "tsconfig.json",
            encoding: "utf-8",
            fileSystemWrapper: fileSystemWrapper
        });
        var compilerOptions = utils_1.getCompilerOptionsFromTsConfigParseResult({
            tsConfigFilePath: "tsconfig.json",
            fileSystemWrapper: fileSystemWrapper,
            tsConfigParseResult: tsConfigParseResult
        });
        var result = utils_1.getPathsFromTsConfigParseResult({
            tsConfigFilePath: "tsconfig.json",
            encoding: "utf-8",
            fileSystemWrapper: fileSystemWrapper,
            tsConfigParseResult: tsConfigParseResult,
            compilerOptions: compilerOptions.options
        });
        return {
            filePaths: result.filePaths.sort(),
            directoryPaths: result.directoryPaths.sort()
        };
    }
    function doTest(fileSystem, expectedPaths) {
        chai_1.expect(getFileAndDirPaths(fileSystem)).to.deep.equal({
            filePaths: expectedPaths.files.sort(),
            directoryPaths: expectedPaths.dirs.sort()
        });
    }
    it("should get the file paths when there are none", function () {
        var fs = new fileSystem_1.VirtualFileSystemHost();
        fs.writeFileSync("tsconfig.json", "{ \"compilerOptions\": { \"rootDir\": \"test\", \"target\": \"ES5\" } }");
        doTest(fs, {
            files: [],
            dirs: ["/test"]
        });
    });
    it("should add the files from tsconfig.json, but exclude the outDir", function () {
        var fs = new fileSystem_1.VirtualFileSystemHost();
        fs.writeFileSync("tsconfig.json", "{ \"compilerOptions\": { \"outDir\": \"test2\" } }");
        fs.writeFileSync("/otherFile.ts", "");
        fs.writeFileSync("/test/file.ts", "");
        fs.writeFileSync("/test2/file2.ts", "");
        doTest(fs, {
            files: ["/otherFile.ts", "/test/file.ts"],
            dirs: ["/", "/test"]
        });
    });
    it("should add the files from tsconfig.json, but exclude the specified exclude", function () {
        var fs = new fileSystem_1.VirtualFileSystemHost();
        fs.writeFileSync("tsconfig.json", "{ \"exclude\": [\"test2\"] }");
        fs.writeFileSync("/otherFile.ts", "");
        fs.writeFileSync("/test/file.ts", "");
        fs.writeFileSync("/test2/file2.ts", "");
        doTest(fs, {
            files: ["/otherFile.ts", "/test/file.ts"],
            dirs: ["/", "/test"]
        });
    });
    it("should add the files from tsconfig.json, but only include the specified include", function () {
        var fs = new fileSystem_1.VirtualFileSystemHost();
        fs.writeFileSync("tsconfig.json", "{ \"include\": [\"test2\"] }");
        fs.writeFileSync("/otherFile.ts", "");
        fs.writeFileSync("/test/file.ts", "");
        fs.writeFileSync("/test2/file2.ts", "");
        doTest(fs, {
            files: ["/test2/file2.ts"],
            dirs: ["/test2"]
        });
    });
    it("should add the files from tsconfig.json when using rootDir", function () {
        var fs = new fileSystem_1.VirtualFileSystemHost();
        fs.writeFileSync("tsconfig.json", "{ \"compilerOptions\": { \"rootDir\": \"test\" } }");
        fs.writeFileSync("/otherFile.ts", "");
        fs.writeFileSync("/test/file.ts", "");
        fs.writeFileSync("/test/test2/file2.ts", "");
        doTest(fs, {
            files: ["/test/file.ts", "/test/test2/file2.ts"],
            dirs: ["/test", "/test/test2"]
        });
    });
    it("should add the files from tsconfig.json when using rootDirs", function () {
        var fs = new fileSystem_1.VirtualFileSystemHost();
        fs.writeFileSync("tsconfig.json", "{ \"compilerOptions\": { \"rootDirs\": [\"/test/test1\", \"/test/test2\"] } }");
        fs.writeFileSync("/test/file.ts", "");
        fs.writeFileSync("/test/test1/file1.ts", "");
        fs.writeFileSync("/test/test2/file2.ts", "");
        fs.writeFileSync("/test/test2/sub/file3.ts", "");
        doTest(fs, {
            files: ["/test/test1/file1.ts", "/test/test2/file2.ts", "/test/test2/sub/file3.ts"],
            dirs: ["/test/test1", "/test/test2", "/test/test2/sub"]
        });
    });
    it("should add the files from tsconfig.json when using rootDir and rootDirs", function () {
        var fs = new fileSystem_1.VirtualFileSystemHost();
        fs.writeFileSync("tsconfig.json", "{ \"compilerOptions\": { \"rootDir\": \"/test/test1\", \"rootDirs\": [\"/test/test2\"] } }");
        fs.writeFileSync("/test/file.ts", "");
        fs.writeFileSync("/test/test1/file1.ts", "");
        fs.writeFileSync("/test/test2/file2.ts", "");
        doTest(fs, {
            files: ["/test/test1/file1.ts", "/test/test2/file2.ts"],
            dirs: ["/test/test1", "/test/test2"]
        });
    });
});
